src/rollup/index.ts

```diff
@@ -124,357 +102,375 @@ function getInputOptions(rawInputOptions: GenericConfigObject): any {
 }
 
 export default function rollup(
-	rawInputOptions: GenericConfigObject | InputOptions
-): Promise<Bundle | BundleSet> {
+	rawInputOptions: GenericConfigObject
+): Promise<RollupSingleFileBuild | RollupBuild> {
 	try {
 		const inputOptions = getInputOptions(rawInputOptions);
 		initialiseTimers(inputOptions);
 		const graph = new Graph(inputOptions);
 
-		timeStart('BUILD', 1);
+		for (let plugin of graph.plugins) {
+			if (plugin.buildStart) plugin.buildStart.call(graph.pluginContext, inputOptions);
```

Comment:
Using `this.warn` in the `buildStart` and `buildEnd` hooks also caused maximum callstack errors. I think we should have a few more simple tests for the plugin context API. There are already some tests in the `function` category for `this.warn` that could probably be easily adapted for the other hooks.